docker/manager.py

```diff
@@ -0,0 +1,821 @@
+        schema:
+          type: string
+      cudnn:
+        type: [string, list]
+        check_with: dists_defined
+        dependencies: runtime
+
+cuda_dependencies:
+  type: dict
+  nullable: false
+  keysrules:
+    type: string
+    regex: '(\d{1,2}\.\d{1}\.\d)?$'
+  valuesrules:
+    type: dict
+    keysrules:
+      type: string
+    valuesrules:
+      type: string
+
+release_spec:
+  required: true
+  type: dict
+  schema:
+    add_to_tags:
+      type: string
+      required: true
+    templates_dir:
+      type: string
+    base_image:
+      type: string
+    header:
+      type: string
+    cuda_prefix_url:
+      type: string
+      dependencies: cuda
+    cuda_requires:
+      type: string
+      dependencies: cuda
+    needs_deps_image:
+      type: boolean
+    envars:
+      type: list
+      default: []
+      schema:
+        check_with: args_format
+        type: string
+    cuda:
+      type: dict
+
+releases:
+  type: dict
+  keysrules:
+    supported_dists: true
+    type: string
+  valuesrules:
+    type: dict
+    keysrules:
+      regex: '([\d\.]+)'
+    valuesrules:
+      type: dict
+    
+"""
+
+
+def validate_template_name(names):
+    """Validate our input file name for Dockerfile templates."""
+    if isinstance(names, str):
+        if REPO_PREFIX in names:
+            # ignore .repo.j2 for centos
+            pass
+        elif DOCKERFILE_PREFIX in names:
+            names, _, _ = names.partition(DOCKERFILE_PREFIX)
+            if names not in DOCKERFILE_ALLOWED_NAMES:
+                raise RuntimeError(
+                    f"Unable to parse {names}. Dockerfile directory should consists of {', '.join(DOCKERFILE_ALLOWED_NAMES)}"
+                )
+        else:
+            raise RuntimeWarning(
+                f"{inspect.currentframe().f_code.co_name} is being used to validate {names}, which has unknown extensions, "
+                f"thus will have no effects. "
+            )
+            pass
+    elif isinstance(names, list):
+        for _name in names:
+            validate_template_name(_name)
+    else:
+        raise TypeError(f"{names} has unknown type {type(names)}")
+
+
+def get_data(obj, *path, skip=False):
+    """
+    Get data from the object by dotted path.
+    e.g: dependencies.cuda."11.3.1"
+    """
+    try:
+        data = glom(obj, Path(*path))
+    except PathAccessError:
+        if skip:
+            return
+        logging.error(
+            f"Exception occurred in get_data, unable to retrieve from {' '.join(*path)}"
+        )
+    else:
+        return data
+
+
+def update_data(obj, value, *path):
+    """
+    Update data from the object with given value.
+    e.g: dependencies.cuda."11.3.1"
+    """
+    try:
+        _ = glom(obj, Assign(Path(*path), value))
+    except PathAssignError:
+        logging.error(
+            f"Exception occurred in update_data, unable to update {Path(*path)} with {value}"
+        )
+
+
+def _build_release_args(bentoml_version):
```

Comment:
We will define our `ENV` with this function. Hardcoded `PYTHON_VERSION` and `BENTOML_VERSION`. Then later update values with build context.